WebsocketProviderV2 updates + formatting middleware bugfix

fselmo opened the pull request:
### What was wrong?

- Closes #3093
- `WebsocketProviderV2` needed tests for subscriptions. While we revamp the test suite and try to get some actual integration tests for `eth_subscribe`, introduce some static tests to make sure our formatters and middleware work appropriately for subscriptions.
- Adds a timeout around trying to match the response `id` to the request `id` for a request. Also improves on the logic around matching the ids, simplifying it.
- Add unit tests for the raw response cache process. 
  - Ensure we are caching any undesired responses before returning the desired response.
  - Ensure we return the desired cached response, if cached, and that we never call recv() in that case. 
  - Test the timeout if a response with matching `id` never comes (see above bullet point for test case).
- **(breaking change for beta provider)** Refactor to a proper AsyncIterator pattern for receiving a persistent message stream from the socket.
- Update docs based on the above

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![20231004_181437](https://github.com/ethereum/web3.py/assets/3532824/a7369729-ecc7-4db7-98f5-87c77d5c042f)

fselmo committed:
Add async_geth_poa_middleware for subscription responses

fselmo committed:
Refactor async iterator pattern for websocket recv()

- Add a proper ``__anext()__`` setup for listening to an open websocket connection. This is the more appropriate pattern without the necessity for a while loop. This also more directly mirrors the ``websockets`` library example and design pattern.
- Update documentation example to reflect the above changes.

fselmo committed:
Add static tests for eth_subscribe while we don't have something better

- Change async_generator -> async_ctx_manager (Context Manager) as this is a more appropriate name for the pattern.
- Create a PersistentConnectionProviderTest class and have both the context manager and iterator test suite run these tests.
- Add static eth_subscribe tests to keep our formatters and middlewares in check.

+ additional improvements to typing; pass linting

fselmo committed:
Add a test for async_geth_poa_middleware w/ subscriptions

fselmo committed:
relevant newsfragments for #3116

fselmo committed:
Minor additions

- Simplify the json encoding for the static eth_subscription tests
- Connect to websocket when entering the AsyncIterator pattern for WebsocketsProviderV2 if it isn't already connected

fselmo committed:
Simplify the async lock around transient caches

- We create transient caches on a per-provider basis. This wouldn't be shared among threads to begin with. The async_lock() utility was designed because the request session cache exists outside of any class and thus is shared across many instances of classes that utilize it. Change the async lock around transient caches to be a standard use of the ``asyncio.Lock()``.
- Move the lock into the RequestProcessor class since that's where the caches live.

fselmo committed:
Improve on inefficiencies with caching logic

- We weren't caching every time we held a response. This lead to missing some messages entirely from the provider. This will need some better testing to ensure this logic never breaks.
- When popping a raw response, actually return it. I'll add a test for this in a separate commit as well.

reedsa commented on docs/providers.rst:
```diff
@@ -270,16 +270,15 @@ asynchronous context manager, can be found in the `websockets connection`_ docs.
+        ...             if some_condition:
+        ...                 # unsubscribe from new block headers and break out of
+        ...                 # iterator
+        ...                 unsubscribed = await w3.eth.unsubscribe(subscription_id)
```
Any reason to set the return value here to `unsubscribed`? Now that the wile loop is gone, perhaps dont need this variable.
```suggestion
        ...                await w3.eth.unsubscribe(subscription_id)
```

fselmo replied:
Good eye 👌🏼

reedsa commented on web3/_utils/module_testing/persistent_connection_provider.py:
```diff
@@ -0,0 +1,325 @@
+        expected_formatted_result: Any,
+    ) -> None:
+        sub_id = await async_w3.eth.subscribe(*subscription_params)
+        assert sub_id is not None
```
Nit: should we keep this assertion or add it to the test below? Looks like `is_hexstr` returns False for None anyway.

fselmo committed:
Add unit tests for WebsocketProviderV2 make request

- Test that we cache any undesired messages until we find the one we want and return it.
- Test that we return the cached response if in cache without ever calling recv()

fselmo committed:
Tweak formatters comparison % match; add syncing tests

- Add tests for ``syncing`` eth_subscribe requests
- Tweak formatters keyset match to 75% match. Across all eth_subscribe requests, the keys are not close to being similar so we can afford a more relaxed check here.

reedsa reviewed (commented):
Remaining nits, looking good otherwise!

fselmo committed:
Changes from comments on PR #3116

fselmo committed:
Add timeout around request / response id matching while loop:

- Use the call timeout for the provider to set up a reasonable escape time for the while look that tries to match the response `id` with the request `id`. All providers should technically return matching ids in the response but we've run into issues with this in the past and this prevents an infinite loop, handling it somewhat gracefully.
- Make sure that the ``WebsocketProviderV2`` sets a default timeout.

reedsa commented on web3/providers/websocket/websocket_v2.py:
```diff
@@ -158,40 +157,52 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
+                f"Timed out waiting for response to request id `{request_id}` after "
+                f"{self.call_timeout} seconds. This is likely due to the provider not "
+                "issuing a response with the same id that was sent in the request, "
+                "which is required by the JSON-RPC spec."
```
```suggestion
                f"Timed out waiting for response with request id `{request_id}` after "
                f"{self.call_timeout} seconds. This is likely due to the provider not "
                "returning a response with the same id that was sent in the request, "
                "which is required by the JSON-RPC spec."
```

reedsa commented on web3/providers/websocket/websocket_v2.py:
```diff
@@ -67,7 +66,7 @@ def __init__(
         endpoint_uri: Optional[Union[URI, str]] = None,
         websocket_kwargs: Optional[Dict[str, Any]] = None,
-        call_timeout: Optional[int] = None,
+        call_timeout: Optional[float] = DEFAULT_PERSISTENT_CONNECTION_TIMEOUT,
```
The timeout may be `None` or could be the max float value, `wait_for` will in turn wait forever (or until the timeout of the large value provided). Since this would be intentionally set by the user, I think it is acceptable to let the connection stay open until the response id is returned, if ever. What do you think?

fselmo replied:
Hmm. That's a good point. We use this not only for call timeouts but for the while loop timeout. What if we set the while loop timeout to `self.call_timeout or 20` so that if it is set to `None` for call timeouts, the while loop will still max out in `20` seconds. Or whatever we decide that default is. Since it differs from the timeout for each call, I think that might be a good alternative. Thoughts?

reedsa replied:
Makes sense to have the timeout limit on the while loop. Paraphrasing what you said to make sure I have it right in my head, use the `call_timeout` unless it exceeds the threshold (20s or something). If `call_timeout` is larger than the threshold, the while loop will still  short circuit as desired.

fselmo replied:
My suggestion would've only worked for `None` but I think yours works well.

```py
timeout = self.call_timeout if self.call_timeout and self.call_timeout <= 20 else 20
```

This also prevents it from being `0` in the `if` check. Anything from `0.1` to `20` is fair game.

pacrob reviewed (approved):
lgtm!

reedsa commented on web3/providers/websocket/websocket_v2.py:
```diff
@@ -158,40 +157,55 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
+            # "bad" cases somewhat gracefully.
+            timeout = (
+                self.call_timeout
+                if self.call_timeout and self.call_timeout <= 20
+                else 20
```
Should use the `DEFAULT_PERSISTENT_CONNECTION_TIMEOUT` constant here.
```suggestion
                if self.call_timeout and self.call_timeout <= DEFAULT_PERSISTENT_CONNECTION_TIMEOUT
                else DEFAULT_PERSISTENT_CONNECTION_TIMEOUT
```

fselmo replied:
Isn't this an entirely different bit of logic though? What if we agree to change the default timeout? I think that was the point of hard-coding this here.

Let's say I don't want _each_ call to the provider to take more than, say, 5 seconds. I wonder if I'm subscribed to a lot of subscriptions, if the response to my call won't be bogged down by the latency behind so many subscriptions where it may actually take longer than a call timeout to receive it.

Thoughts there?

reedsa replied:
That's fine then, I have no concerns with using the hardcoded value as is. Definitely would be overkill to create yet another constant just for this logic.

fselmo replied:
Yeah in the same way I think using `self.call_timeout` feels a bit disingenuous to use here too but at least the user can tweak that to make sense if desired. Definitely open to re-thinking this a bit more before ripping off the `beta` tag.

reedsa reviewed (approved):
🚀 :shipit:
